### PR TITLE
Ux bootstrap revamp 20210706

### PR DESCRIPTION
**Pull request by aahung**
#### Checklist

- [ ] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [ ] Write unit tests
- [ ] Write/update functional tests
- [ ] Write/update integration tests
- [ ] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

File: samcli/commands/pipeline/bootstrap/guided_context.py [hunk 1]
```diff
@@ -33,81 +36,195 @@ def __init__(
         self.pipeline_ip_range = pipeline_ip_range
         self.region = region
 
+    def _prompt_stage_name(self) -> None:
+        click.echo(
+            "Enter a name for the stage you want to bootstrap. This will be referenced later "
```
Comment (c2tarun) [inline samcli/commands/pipeline/bootstrap/guided_context.py:41]: `Enter a name` -> `Enter name`

File: samcli/commands/pipeline/bootstrap/guided_context.py [hunk 2]
```diff
@@ -33,81 +36,195 @@ def __init__(
...
+        )
+        self.environment_name = click.prompt(
+            "Stage name",
+            default=self.environment_name,
+            type=click.STRING,
+        )
+
+    def _prompt_region_name(self) -> None:
+        self.region = click.prompt(
+            "Enter the region you want these resources to create",
```
Comment (c2tarun) [inline samcli/commands/pipeline/bootstrap/guided_context.py:52]: Do we need to clarify what `these` mean?

**Review by c2tarun — APPROVED** [review]

File: samcli/commands/pipeline/bootstrap/guided_context.py [hunk 3]
```diff
@@ -33,81 +36,195 @@ def __init__(
         self.pipeline_ip_range = pipeline_ip_range
         self.region = region
 
+    def _prompt_stage_name(self) -> None:
+        click.echo(
+            "Enter a name for the stage you want to bootstrap. This will be referenced later "
+            "when generating a Pipeline Config File with Pipeline Init."
```
Comment (hoffa) [inline samcli/commands/pipeline/bootstrap/guided_context.py:42]: We should only capitalize proper nouns. Are we using "config" for prompts versus "configuration"? Not sure whether we want to add backticks to CLI commands.

```suggestion
            "when generating a pipeline configuration file with pipeline init."
```

File: samcli/commands/pipeline/bootstrap/guided_context.py [hunk 4]
```diff
@@ -33,81 +36,195 @@ def __init__(
...
+                text="Press enter to confirm the values above, or select an item to edit the value",
+                default="0",
+                show_choices=False,
+                show_default=False,
+                type=click.Choice(["0"] + [str(i + 1) for i in range(len(inputs))]),
             )
+            if int(edit_input):
+                inputs[int(edit_input) - 1][1]()
+            else:
+                break
```
Comment (hoffa) [inline samcli/commands/pipeline/bootstrap/guided_context.py:230]: Shouldn't the input by this point be _guaranteed_ valid? i.e. `click` should force a valid response.

File: samcli/commands/pipeline/external_links.py [hunk 5]
```diff
@@ -0,0 +1,5 @@
+"""
```
Comment (c2tarun) [inline samcli/commands/pipeline/external_links.py:1]: Just wanted to confirm that we are going to add integ test for validating broken link later.

File: samcli/commands/pipeline/bootstrap/cli.py [hunk 6]
```diff
@@ -206,14 +207,26 @@ def do_cli(
...
-            Colored().green(
-                "\nThe ARNs of created resources have been written to "
-                f"{os.path.join(PIPELINE_CONFIG_DIR, PIPELINE_CONFIG_FILENAME)}.\n"
-                f"It will be used next time you run `sam pipeline bootstrap` or "
-                f"`sam pipeline init` in this directory."
+            dedent(
+                f"""\
+                View the definition in {os.path.join(PIPELINE_CONFIG_DIR, PIPELINE_CONFIG_FILENAME)},
+                run {Colored().bold("sam pipeline bootstrap")} to generate another set of resources, or proceed to
+                {Colored().bold("sam pipeline init")} to create your Pipeline Config file.
```
Comment (hoffa) [inline samcli/commands/pipeline/bootstrap/cli.py:214]:
```suggestion
                {Colored().bold("sam pipeline init")} to create your pipeline configuration file.
```

File: samcli/commands/pipeline/bootstrap/guided_context.py [hunk 7]
```diff
@@ -32,82 +37,198 @@ def __init__(
...
+    def _prompt_region_name(self) -> None:
+        self.region = click.prompt(
+            "Enter the region you want these resources to create",
+            type=click.STRING,
+            default=get_default_aws_region(),
+        )
+
+    def _prompt_pipeline_user(self) -> None:
+        self.pipeline_user_arn = click.prompt(
+            "Enter the Pipeline IAM User ARN if you have previously created one, or we will create one for you",
```
Comment (hoffa) [inline samcli/commands/pipeline/bootstrap/guided_context.py:62]:
```suggestion
            "Enter the pipeline IAM user ARN if you have previously created one, or we will create one for you",
```

File: samcli/commands/pipeline/bootstrap/guided_context.py [hunk 8]
```diff
@@ -32,82 +37,198 @@ def __init__(
...
+            default="",
+            type=click.STRING,
+        )
+
+    def _prompt_pipeline_execution_role(self) -> None:
+        self.pipeline_execution_role_arn = click.prompt(
+            "Enter the Pipeline execution role ARN if you have previously created one, "
```
Comment (hoffa) [inline samcli/commands/pipeline/bootstrap/guided_context.py:69]:
```suggestion
            "Enter the pipeline execution role ARN if you have previously created one, "
```

File: samcli/commands/pipeline/bootstrap/guided_context.py [hunk 9]
```diff
@@ -32,82 +37,198 @@ def __init__(
...
+            "Please enter the artifact bucket ARN for your Lambda function. "
+            "If you do not have a bucket, we will create one for you",
+            default="",
+            type=click.STRING,
+        )
+
+    def _prompt_image_repository(self) -> None:
+        if click.confirm("Does your application contain any IMAGE type Lambda functions?"):
+            self.image_repository_arn = click.prompt(
+                "Please enter the ECR image repository ARN(s) for your IMAGE type function(s)."
```
Comment (hoffa) [inline samcli/commands/pipeline/bootstrap/guided_context.py:94]:
```suggestion
                "Please enter the ECR image repository ARN(s) for your Image type function(s)."
```

File: samcli/commands/pipeline/bootstrap/guided_context.py [hunk 10]
```diff
@@ -32,82 +37,198 @@ def __init__(
...
+                default="",
+                type=click.STRING,
+            )
+            self.create_image_repository = not bool(self.image_repository_arn)
+        else:
+            self.create_image_repository = False
+
+    def _prompt_ip_range(self) -> None:
+        self.pipeline_ip_range = click.prompt(
+            "For added security, you can define the permitted Pipeline IP range. "
```
Comment (hoffa) [inline samcli/commands/pipeline/bootstrap/guided_context.py:105]:
```suggestion
            "For added security, you can define the permitted pipeline IP range. "
```

File: samcli/commands/pipeline/bootstrap/guided_context.py [hunk 11]
```diff
@@ -32,82 +37,198 @@ def __init__(
...
+    def run(self) -> None:  # pylint: disable=too-many-branches
         """
         Runs an interactive questionnaire to prompt the user for the ARNs of the AWS resources(infrastructure) required
         for the pipeline to work. Users can provide all, none or some resources' ARNs and leave the remaining empty
         and it will be created by the bootstrap command
         """
+        click.secho(
+            dedent(
+                """\
+                SAM Pipeline Bootstrap generates the necessary AWS resources to connect your
```
Comment (hoffa) [inline samcli/commands/pipeline/bootstrap/guided_context.py:162]:
```suggestion
                sam pipeline bootstrap generates the necessary AWS resources to connect your
```

File: samcli/commands/pipeline/bootstrap/guided_context.py [hunk 12]
```diff
@@ -32,82 +37,198 @@ def __init__(
...
+                CI/CD pipeline tool. We will ask for [1] account details, [2] stage definition,
```
Comment (hoffa) [inline samcli/commands/pipeline/bootstrap/guided_context.py:163]:
```suggestion
                CI/CD system. We will ask for [1] account details, [2] stage definition,
```

File: samcli/commands/pipeline/bootstrap/guided_context.py [hunk 13]
```diff
@@ -32,82 +37,198 @@ def __init__(
...
+
         account_id = get_current_account_id()
-        if not self.environment_name:
-            self.environment_name = click.prompt(
-                f"Environment name (a descriptive name for the environment which will be deployed"
-                f" to AWS account {account_id})",
-                type=click.STRING,
-            )
+        click.secho("[1] Account details", bold=True)
+        if click.confirm(f"You are bootstrapping resources in Account {account_id}. Do you want to switch accounts?"):
```
Comment (hoffa) [inline samcli/commands/pipeline/bootstrap/guided_context.py:173]:
```suggestion
        if click.confirm(f"You are bootstrapping resources in account {account_id}. Do you want to switch accounts?"):
```

File: samcli/commands/pipeline/bootstrap/guided_context.py [hunk 14]
```diff
@@ -32,82 +37,198 @@ def __init__(
...
-                "If you have ran this command before to bootstrap a previous environment, please "
-                "provide the ARN of the created pipeline user, otherwise, we will create a new user for you. "
-                "Please make sure to store the credentials safely with the CI/CD system."
-            )
-            self.pipeline_user_arn = click.prompt(
-                "Pipeline user [leave blank to create one]", default="", type=click.STRING
-            )
+        click.secho("[3] Reference existing resources", bold=True)
+        if self.pipeline_user_arn:
+            click.echo(f"Pipeline IAM User ARN: {self.pipeline_user_arn}")
```
Comment (hoffa) [inline samcli/commands/pipeline/bootstrap/guided_context.py:188]:
```suggestion
            click.echo(f"Pipeline IAM user ARN: {self.pipeline_user_arn}")
```

**Review by hoffa — APPROVED** [review]

**Pull request merged**